tests/cancun/eip6780_selfdestruct/test_create2_collision_selfdestruct.py

```diff
@@ -0,0 +1,112 @@
+"""
+
+import pytest
+
+from ethereum_test_forks import Fork
+from ethereum_test_tools import (
+    Account,
+    Environment,
+    Initcode,
+    StateTestFiller,
+    TestAddress,
+    TestAddress2,
+    Transaction,
+    to_address,
+)
+from ethereum_test_tools.vm.opcode import Opcodes as Op
+
+REFERENCE_SPEC_GIT_PATH = "EIPS/eip-6780.md"
+REFERENCE_SPEC_VERSION = "2f8299df31bb8173618901a03a8366a3183479b0"
+
+
+@pytest.fixture
+def env():  # noqa: D103
+    return Environment(
+        coinbase="0x2adc25665018aa1fe0e6bc666dac8fc2697ff9ba",
+        difficulty=0x020000,
+        gas_limit=71794957647893862,
+        number=1,
+        timestamp=1000,
+    )
+
+
+@pytest.mark.valid_from("Merge")
+def test_create2_collision_selfdestruct(
+    env: Environment,
+    fork: Fork,
+    state_test: StateTestFiller,
+):
+    """Dynamic Create2->Suicide->Create2 collision scenario:"""
+    """
+    Perform a CREATE2, make sure that the initcode sets at least a couple of storage keys,
+    then on a different call, in the same tx, perform a self-destruct.
+    Then:
+        a) on the same tx, attempt to recreate the contract   <=== Covered int this test
+        b) on a different tx, attempt to recreate the contract
+    Perform a CREATE2, make sure that the initcode sets at least a couple of storage keys,
+    then in a different tx, perform a self-destruct.
+    Then:
+        a) on the same tx, attempt to recreate the contract
+        b) on a different tx, attempt to recreate the contract
+    Verify that the test case described in https://wiki.hyperledger.org/pages/viewpage.action?pageId=117440824 is covered
+    """
+
+    address_to = TestAddress2
+    address_s = to_address(0x0512)
+    suicide_d = to_address(0x03E8)
+    create2_address = to_address(0x15A1593FA2C3A4FAA0DD9109C64860E906ECD6)
+
+    pre = {
+        address_to: Account(
+            balance=1000000000000000000,
+            nonce=0,
+            code=Op.CALLDATACOPY(0, 0, Op.CALLDATASIZE())
+            + Op.MSTORE(0, Op.CREATE2(0, 0, Op.CALLDATASIZE(), 1))
+            + Op.CALL(Op.GAS(), Op.MLOAD(0), 0, 0, 0, 0, 0)
+            + Op.CALLDATACOPY(0, 0, Op.CALLDATASIZE())
+            + Op.SSTORE(1, Op.CREATE2(0, 0, Op.CALLDATASIZE(), 1)),
+            storage={0x01: 0xFF},
+        ),
+        address_s: Account(
+            balance=7000000000000000000,
+            nonce=0,
+            code=Op.SSTORE(1, 1),
+            storage={},
+        ),
+        TestAddress: Account(
+            balance=7000000000000000000,
+            nonce=0,
+            code="0x",
+            storage={},
+        ),
+    }
+
+    post = {
+        # Make sure the CREATE2 address performed storage operations
+        address_s: Account(storage={0x01: 0x01}),
+        # Make sure the second CREATE2 attempt is failing
+        address_to: Account(storage={0x01: 0x00}),
+        # Make sure selfdestruct cleans the account
+        create2_address: Account.NONEXISTENT,
```

Comment:
We could parametrize another flavor of this test where the `create2_address` is already created before this block, and we test that even though create2 was called in this same tx, the self-destruct later is still inhibited.

As per @MariusVanDerWijden, there could be a bug in a client where they set a `create2` flag somewhere when the opcode is called, and then they could mistakenly try to delete the account even though it was created before the tx.